CEL based v2 attribute generation: Request Classification.

mandarjog opened the pull request:
Add: proto definition for request classification.
Add: two examples for use.
Add: Implementation
Add: Unit tests.

Note: the configuration must be json at the point of delivery, not yaml.


An Alternate to #2799

istio-testing commented:
@mandarjog: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | 0731349e48d194128590a36249e3b61f60e5a956 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/586) | `/test test_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nrjpoddar commented on extensions/classify/config.proto:
```diff
@@ -0,0 +1,98 @@
+  // This is the correct behaviour if the result of the classification is used by an authz plugin.
+  // If the result of the classifier is used by stats, you may just want to log an error but not
+  // fail the request. If `ignore_errors=true` then the classifier plugin will not fail the request.
+  bool ignore_errors = 3;
```
This is interesting and can have broad security implications if any security features are dependent on these attributes.

Should this be always set to "false" and not even exposed as a field i.e. "attribute" is best effort with defaults. Downstream filters should be created in a way that they handle unset/default/actual value.

douglas-reid replied:
If possible, i'd prefer to have some distinction between failure and unset. maybe `istio.attributeGen.status` should be an attribute ;p.

mandarjog replied:
This exactly gives us the difference between failure to evaluate and evaluated to empty.
We could put the onus of failing the request of the consumer of the attribute and not the producer.

nrjpoddar commented on extensions/classify/config.proto:
```diff
@@ -0,0 +1,98 @@
+  string condition = 1;
+
+  // If condition evaluates to true, return the `value`.
+  string value = 2;
```
should we have a concept of "default"?

mandarjog replied:
default is just a `Match clause` with no condition.
I had it in the other PR. forgot to copy it.

nrjpoddar commented on extensions/classify/config.proto:
```diff
@@ -0,0 +1,98 @@
+
+// Classification uses [well known attributes](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/rbac_filter#condition) as inputs and produces
+// a new attribute that denotes a specific classification.
+// The new attribute can be used by a downstream plugin.
```
plugin vs filter which is preferable?

mandarjog replied:
We call it plugin in the Wasm paralance, so I prefer plugin.

nrjpoddar commented:
This is basically bringing Mixer back in a much more performant way (hopefully). :)

nrjpoddar commented on extensions/classify/config.proto:
```diff
@@ -0,0 +1,98 @@
+message PluginConfig {
+  // next id: 2
+  // Multiple independent classification configurations.
+  repeated Classification classifications = 1;
```
We still need to add documentation around the behavior of overlapping "output_attribute" i.e. what happens if the config has `operation` twice. Does the last one win?

nrjpoddar replied:
Classify is too narrow a use case for this implementation since this can be used in an abstract way.

Other name suggestion:
repeated AttributeConfig attributes = 1;

douglas-reid replied:
`AttributeGeneration` would match the v1 name for these sorts of extensions. Not sure if that helps or not.

mandarjog replied:
I agree classification is too narrow, this is now an APA from v1.

douglas-reid commented on extensions/classify/config.proto:
```diff
@@ -0,0 +1,98 @@
+
+syntax = "proto3";
+
+package istioclassify;
```
any reason not to make this `istio.classify` ?  or `istio.attributegen` ?

mandarjog replied:
istio.attributegen sounds good.

douglas-reid commented on extensions/classify/config.proto:
```diff
@@ -0,0 +1,98 @@
+
+  // A Classification may fail to evaluate when an attribute is not available.
+  // Example: `response.code` may not be available when a request ends abruptly.
+  // When evalution failes, the plugin will return an error to the client.
```
```suggestion
  // When evalution fails, the plugin will return an error to the client.
```

mandarjog replied:
ack

douglas-reid commented on extensions/classify/config.proto:
```diff
@@ -0,0 +1,98 @@
+
+  // The following attribute is populated on successfully running the classification.
+  // Example: `istio.operationId`
+  string output_attribute = 2;
```
do we have guidance on attribute namespaces?  like `istio.*` is reserved ?

istio-testing commented:
@mandarjog: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | ef0188f10d8c17f4f451da25cf2283e2a21cb1a1 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/587) | `/test test_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | 4f0f8c5a74f8bce1180bcdae3bdd9f368b0be9a7 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/588) | `/test test_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,113 @@
+  string condition = 1;
+
+  // If condition evaluates to true, return the `value`.
+  string value = 2;
```
At this time, value is a constant.
Should also add output_expr or wait until we need it?

nrjpoddar replied:
I would add it later if requested.

douglas-reid replied:
If we are going to expand it later, should this be:

`Value value = 2;`

Using a message would allow us flexibility.

mandarjog replied:
It adds to the level of indentation :-( but these are not supposed to be written by a human !

mandarjog replied:
could be do a one_of (value and value_expr)?

nrjpoddar commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,113 @@
+
+package istio.attributegen;
+
+// Configuration to generate attributes based on existing attributes.
```
Nit: This is a bit difficult to understand, suggestion: 

Configuration to generate synthetic attributes based on attributes of the proxied traffic

Something like this might be more clear.

nrjpoddar commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,113 @@
+  // All information about the request is available in this phase,
+  // and classification does not add to the latency observed by the client.
+  PHASE_UNSPECIFIED = 0;
+  AFTER_RESPONSE = 0;
```
Using response/request is fine for now and will work for HTTP, how would you name it so that’s it’s suitable for other protocols?

I have used “transaction” as an abstract way to define this in the past but Istio doesn’t really use transaction in the current vocabulary.

mandarjog replied:
REQUEST, REPLY and LOG can be mapped to most other protocols.

1. memcache / redis all 3 phases are clear
2. mysql  same.
3.kafka  REQUEST (Message). and LOG are ok. There is no reply for payload.
4. grpc stream is like kafka, and bidi-grpc is like 2 kafkas... 

can you suggest something else? @kyessenov  ?

nrjpoddar replied:
Can we remove this option for now since we only support one callback insertion point.
I just think it's too much for a user to configure both filter & callback placements. 

Long term I was thinking for filters & extensions we should defined "classes/categories" and based on the class your placement in filter chain is more or less. This same category can also dictate the phase.

mandarjog replied:
The code supports both insertion points, it is just hidden from the docs.
I am ok hiding Phase completely if you want, but if you need to classify and use the attribute for policy decisions the *must* specify the callback not just the insertion order.

We can recast phases as filter `functions`, as it is done in the extensions API to make it less implementation specific.
This does determine the plugin ordering, but the plugin also needs to know where it will run.

```proto
enum Function {
    UNSPECIFIED_FUNCTION = 0;
    TELEMETRY = 1;
    PRE_PROCESSING = 2;
    AUTHORIZATION = 3;
    POST_PROCESSING = 4;
}
```

nrjpoddar replied:
Yes! These names make much more sense to an end user. With some documentation around these terms like Pre processing and post processing and examples I think it’s very reasonable and good UX.

I don’t mean to be a bottleneck on naming but I believe getting it right now will save us a lot of headaches in the future.

@dougreid thoughts?

kyessenov replied:
Unfortunately, this categorization is rather simplistic. Even the native filter mix and match functions in the same filters (filter does auth, telemetry, and pre-processing at the same time). It's a lot of work to try to re-organize existing native extensions to follow this paradigm, so I'm not sure we want to commit to that in this filter. I think we want this low-level be as close to the metal as to what actually happens in Envoy, and add some sugar on top if that irks some people the wrong way.

kyessenov replied:
Put it the other way - I don't dismiss UX, but it should reflect the reality :)

mandarjog replied:
Being close to the implementation was my goal initially since this is a filter and xds is not supposed to human consumable or producible. 

However we are caught in the middle because the xds APi *is* exposed as is to the end user for now and so UX discussion comes in.
Having a higher level abstraction like TELEMETRY, et al without the higher level extension mechanism to put things in the correct place only gets us half way.

We have 2 options
1. Stick with what we have since it maps cleanly to envoy. We can de emphasize the phase is the doc and examples.
2. Remove the option completely. but we will be in the same spot as today when someone asks for the feature in the future.

nrjpoddar replied:
My suggestion here would be to hide it until someone asks for it, or we have more understanding of requirements.

@kyessenov I agree about pre-mature APIs, hence why expose something that we are not sure of for now.

mandarjog replied:
Ok done, I hid the option from the proto.

douglas-reid commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,113 @@
+// the `operationId` in the OpenAPI Spec.
+// The Stats plugin will run after this plugin and use `istio.operationId` to populate a dimension.
+//
+// If multiple AttributeGenerations produce the same attribute, the result of the last generation
```
Using last is fine, but I recall that in Mixer once an attribute had a value, any subsequent updates were ignored.

douglas-reid commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,113 @@
+  // AttributeGeneration may fail to evaluate when an attribute is not available.
+  // Example: `response.code` may not be available when a request ends abruptly.
+  // When evaluation fails the plugin will set an error_attribute to the error string.
+  // error_attribute = ${output_attribute}_error
```
interesting. if i have multiple generators that each produce the same `output_attribute` and only one fails, do i see both an error attribute and a usable attribute ?

istio-testing commented:
@mandarjog: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | 471e64f5c637806a4dfe8e36a7253fec9605b414 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/596) | `/test test_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | 471e64f5c637806a4dfe8e36a7253fec9605b414 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/596) | `/test test_proxy`
check-wasm_proxy | 471e64f5c637806a4dfe8e36a7253fec9605b414 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/428) | `/test check-wasm_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | 471e64f5c637806a4dfe8e36a7253fec9605b414 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/596) | `/test test_proxy`
check-wasm_proxy | 7abf22c06192b6daa60283104bbb87e1444e8588 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/431) | `/test check-wasm_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
check-wasm_proxy | 7abf22c06192b6daa60283104bbb87e1444e8588 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/431) | `/test check-wasm_proxy`
test_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/613) | `/test test_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/613) | `/test test_proxy`
check-wasm_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/446) | `/test check-wasm_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/613) | `/test test_proxy`
check-wasm_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/446) | `/test check-wasm_proxy`
test-tsan_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/669) | `/test test-tsan_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/613) | `/test test_proxy`
check-wasm_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/446) | `/test check-wasm_proxy`
test-tsan_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/669) | `/test test-tsan_proxy`
test-asan_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-asan_proxy/627) | `/test test-asan_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
check-wasm_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/446) | `/test check-wasm_proxy`
test-tsan_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/669) | `/test test-tsan_proxy`
test-asan_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-asan_proxy/627) | `/test test-asan_proxy`
test_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/619) | `/test test_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test-tsan_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/669) | `/test test-tsan_proxy`
test-asan_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-asan_proxy/627) | `/test test-asan_proxy`
test_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/619) | `/test test_proxy`
check-wasm_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/452) | `/test check-wasm_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test-asan_proxy | e493f07597cbd041de6517fbf884ecfea617d47b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-asan_proxy/627) | `/test test-asan_proxy`
test_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/619) | `/test test_proxy`
check-wasm_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/452) | `/test check-wasm_proxy`
test-tsan_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/675) | `/test test-tsan_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/619) | `/test test_proxy`
check-wasm_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/452) | `/test check-wasm_proxy`
test-tsan_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/675) | `/test test-tsan_proxy`
test-asan_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-asan_proxy/633) | `/test test-asan_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/619) | `/test test_proxy`
test-tsan_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/675) | `/test test-tsan_proxy`
test-asan_proxy | ea530aeb12b1f2a6f1d15e52ef0e8a0df11d22fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-asan_proxy/633) | `/test test-asan_proxy`
check-wasm_proxy | c0f1be657aeaca6c1cfc6b6d72c2167f52e7d41f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/453) | `/test check-wasm_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented:
@kyessenov @bianpengyuan may I get a review on this?
I will add a few more tests with more matching rules, but otherwise this is ready for review.

mandarjog commented on extensions/attributegen/plugin_test.cc:
```diff
@@ -0,0 +1,423 @@
+                 "ListBooks");
+}
+
+// CEL is unable to get access to response status code.
```
@kyessenov @jplevyak CEL is unable to get access to `response.code` .
This works fine in an actual full manual verification, but does not work in this test.
Setting response header `{":status", "404"}` is not populating response.code internally.

istio-testing commented:
@mandarjog: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
check-wasm_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/457) | `/test check-wasm_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
check-wasm_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/457) | `/test check-wasm_proxy`
test-tsan_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/680) | `/test test-tsan_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
check-wasm_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/457) | `/test check-wasm_proxy`
test-tsan_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/680) | `/test test-tsan_proxy`
test_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/624) | `/test test_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
check-wasm_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/check-wasm_proxy/457) | `/test check-wasm_proxy`
test-tsan_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/680) | `/test test-tsan_proxy`
test_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/624) | `/test test_proxy`
test-asan_proxy | d6e357792605a44d768bea93b294dcc01449f6b0 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-asan_proxy/638) | `/test test-asan_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,124 @@
+// dimension.
+//
+// If multiple AttributeGenerations produce the same attribute, the result of
+// the last generation will be visible to downstream filters.
```
The primary attributes *cannot* be overridden at this time.

The synthetic attributes are set in the filterState and those *can* be overridden.

mandarjog commented:
@nrjpoddar I just went through you comments in the previous PR again, they were already factored in. Please point me if I have missed something.

1. I have hidden phase from the proto.
2. I think `Value value` will create one more level of indent, and we can address it by a one_of later.

istio-testing commented:
@mandarjog: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test-tsan_proxy | f89ac39f1225fa30d3a19126e02c6366a0222a9c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-tsan_proxy/695) | `/test test-tsan_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented:
/test test-tsan_proxy

douglas-reid commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,125 @@
+// value.
+message Match {
+  // Condition is a [CEL
+  // expressions](https://github.com/google/cel-spec/blob/master/doc/langdef.md)
```
```suggestion
  // expression](https://github.com/google/cel-spec/blob/master/doc/langdef.md)
```

douglas-reid commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,125 @@
+  // Note: CEL uses [re2](https://github.com/google/re2/wiki/Syntax) regex
+  // library.
+  //       Use anchors {^, $} to ensure that the regex evaluates efficiently.
+  // Note: request.url_path is normalized and stripped of query params.
```
```suggestion
  // Note: `request.url_path` is normalized and stripped of query params.
```

mandarjog replied:
ack

douglas-reid commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,125 @@
+  // Examples:
+  //
+  // a GET on /books
+  // { value: ListBooks,
```
should these be in pre-formatted blocks?

mandarjog replied:
ack.

douglas-reid commented on extensions/attributegen/config.proto:
```diff
@@ -0,0 +1,125 @@
+}
+
+// Phase determines when the classification is performed.
+enum Phase {
```
i can't recall. does the entire message need to be hidden from docs too?

mandarjog replied:
yes the entire message is hidden.

mandarjog commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,147 @@
+    if (!eval_status) {
+      LOG_DEBUG(absl::StrCat("Failed eval. setting ", ag.error_attribute()));
+      // eval failed set error attribute
+      setFilterState(ag.error_attribute(), "1");
```
Good point about `_error` attribute overshadowing an attribute that is already there.
However since this is part of the contract and API, use will hopefully see this.

We always *need* to filterstate or error so that downstream plugins can distinguish between an error and no-match.

kyessenov replied:
It's unclear why we need error filter state. Filter state is expensive, and just putting "1" there is wasteful.

mandarjog replied:
Please suggest something else that will help the consumer of this attribute distinguish between attribute not set because condition did not match, and attribute could not be evaluated.
The consuming plugin may be authz and it is important to distinguish between the two.

kyessenov replied:
If it's not set, then it's not set. The policy would just evaluate to an error, or use presence the checks for extra control in rare cases.

mandarjog replied:
it is no more wasteful than setting the attribute itself.

kyessenov commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,147 @@
+  // TODO(mjog) see if this still needed since onDone is always called.
+  cleanupAttributeGen();
+
+  for (auto agconfig : config.attributes()) {
```
nit: C++ style is to use long name, attribute_gen_config. This is not golang.

kyessenov commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,147 @@
+      phase = on_request;
+    }
+    auto ag = AttributeGenerator(phase, agconfig.output_attribute());
+    for (auto matchconfig : agconfig.match()) {
```
const auto&

kyessenov commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,147 @@
+        LOG_WARN(absl::StrCat("Cannot create expression: <",
+                              matchconfig.condition(), "> for ",
+                              agconfig.output_attribute()));
+        return false;
```
why fail fast and not ignore?

mandarjog replied:
I do not follow, what do you want to ignore?

kyessenov replied:
Generally we never want to reject config unless something terrible happens since it immediately crashes envoy in some cases. Seems like one bad expression is not a sufficient reason to crash.

kyessenov commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,147 @@
+    std::string val;
+    auto eval_status = ag.evaluate(&val);
+    if (!eval_status) {
+      LOG_DEBUG(absl::StrCat("Failed eval. setting ", ag.error_attribute()));
```
This is noisy. Maybe use trace logging for per-request logging.

kyessenov commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+    }
+  }
+
+ protected:
```
why protected? there's nothing virtual here.

kyessenov commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+      return {};
+    }
+    if (matched) {
+      LOG_DEBUG(absl::StrCat("Matched: ", condition_));
```
Too noisy.

mandarjog replied:
protected all logs under `if (debug_)` or trace.

kyessenov commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+
+  std::string value() const { return value_; };
+
+  void cleanup() {
```
+1

mandarjog replied:
only rootContext has cleanup now.
There was no need for destructors.

kyessenov commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+  EvalPhase phase() const { return phase_; }
+  std::string output_attribute() const { return output_attribute_; }
+  std::string error_attribute() const { return error_attribute_; }
+  void add_match(Match match) { matches_.push_back(match); }
```
Why does this need to be mutable? It's created once and const after that.

mandarjog replied:
removed.

kyessenov commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+  explicit AttributeGenerator(EvalPhase phase, std::string output_attribute)
+      : phase_(phase),
+        output_attribute_(output_attribute),
+        error_attribute_(absl::StrCat(output_attribute, "_error")) {}
```
uninitialized member `matches_`

mandarjog replied:
fixed.

kyessenov commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+ public:
+  PluginRootContext(uint32_t id, StringView root_id)
+      : RootContext(id, root_id) {}
+  ~PluginRootContext() = default;
```
this is tautological?

mandarjog replied:
removed.

kyessenov commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+  void onLog() override { rootContext()->attributeGen(on_log); };
+
+  FilterHeadersStatus onRequestHeaders(uint32_t) override {
+    rootContext()->attributeGen(on_request);
```
There's too much code in the header.  Please move some code into .cc file. That will make imports smaller, too.

mandarjog replied:
moved to .cc

kyessenov commented on extensions/attributegen/plugin_test.cc:
```diff
@@ -0,0 +1,501 @@
+}
+
+TEST_P(AttributeGenFilterTest, OperationFileListNoMatch) {
+  const std::string attribute = "istio.operationId";
```
move into a compilation unit global

kyessenov commented on extensions/attributegen/plugin_test.cc:
```diff
@@ -0,0 +1,501 @@
+  Http::TestRequestHeaderMapImpl request_headers{{":path", "/books"},
+                                                 {":method", "GET"}};
+  Http::TestResponseHeaderMapImpl response_headers{{":status", "504"}};
+  verify_request(request_headers, response_headers, attribute, true, false,
```
verityRequest per style guide

kyessenov commented on extensions/attributegen/testdata/server.yaml:
```diff
@@ -0,0 +1,236 @@
+node:
+  id: test-server-ratings
```
Please use the test framework to add an integration test. I don't really trust the mocking framework. That would help save on duplication of the JSONs in this repo, since you can use one of the templates.

gargnupur commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,147 @@
+      continue;
+    }
+    LOG_DEBUG(absl::StrCat("Setting ", ag.output_attribute(), " --> ", val));
+    setFilterState(ag.output_attribute(), val);
```
I think default span for filterstate is "filterchain", for this plugin, shouldn't it be longer "downstream connection"?
(Not sure if it's available in ABI to do that yet.. would need access to WasmStatePrototype I think)

mandarjog replied:
Good question, @kyessenov  @jplevyak ?

kyessenov replied:
Yes, it's per filter chain (individual pass per request) if you don't declare the property. The ABI is available, see #2817 .

gargnupur commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,147 @@
+      JsonStringToMessage(configuration->toString(), &config, json_options);
+  if (status != Status::OK) {
+    LOG_WARN(absl::StrCat(
+        "Cannot parse plugin configuration JSON string [YAML not supported]: ",
```
nit: add name of this plugin in error message?

gargnupur commented on extensions/attributegen/plugin_test.cc:
```diff
@@ -0,0 +1,501 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
```
nit:2020

mandarjog commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+
+class Match {
+ public:
+  explicit Match(std::string condition, uint32_t condition_token,
```
ack.

mandarjog commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+    return matched;
+  };
+
+  std::string value() const { return value_; };
```
ack.

mandarjog commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+  std::string value_;
+};
+
+enum EvalPhase { on_log, on_request };
```
ack.

mandarjog commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+  }
+
+  EvalPhase phase() const { return phase_; }
+  std::string output_attribute() const { return output_attribute_; }
```
done.

mandarjog commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,205 @@
+  };
+  void attributeGen(EvalPhase);
+
+ protected:
```
private now.

istio-testing commented:
@mandarjog: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | e403c382e37a15bc755398176bce59809b95df8f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/663) | `/test test_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mandarjog: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_proxy | e403c382e37a15bc755398176bce59809b95df8f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test_proxy/663) | `/test test_proxy`
test-asan_proxy | e403c382e37a15bc755398176bce59809b95df8f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2801/test-asan_proxy/675) | `/test test-asan_proxy`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented:
@kyessenov @bianpengyuan @gargnupur I have addressed code review comments.
There are unit tests. I don't want to add integration tests to the same PR and make it bigger.

kyessenov commented on extensions/attributegen/Makefile:
```diff
@@ -0,0 +1,17 @@
+	protoc config.proto --cpp_out=.
+	em++ -s STANDALONE_WASM=1 -s EMIT_EMSCRIPTEN_METADATA=1 -s EXPORTED_FUNCTIONS=['_malloc'] --std=c++17 -O3 -I${CPP_API} -I${CPP_API}/google/protobuf -Iextensions/common -I. -I/usr/local/include -I${ABSL} -I. --js-library ${CPP_API}/proxy_wasm_intrinsics.js $*.cc ${CPP_API}/proxy_wasm_intrinsics.pb.cc ${PROTO_SRCS} ${COMMON_SRCS} ${CPP_CONTEXT_LIB} ${ABSL_CPP} ${CPP_API}/libprotobuf.a -o $*.wasm
+	rm -f $*.wast
+	rm -f extensions/attributegen/config.pb.*
```
rm -f config.pb.*

kyessenov commented on extensions/attributegen/plugin.h:
```diff
@@ -0,0 +1,172 @@
+ public:
+  explicit AttributeGenerator(EvalPhase phase,
+                              const std::string& output_attribute,
+                              const std::vector<Match> matches)
```
`const std::vector<Match>&` or `&&`.

kyessenov commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,233 @@
+      &out_size);
+  if (result != WasmResult::Ok) {
+    LOG_TRACE(absl::StrCat("Failed to evaluate expression:[", condition_token_,
+                           "] ", condition_, " result: ", toString(result)));
```
if out not null, free out.

kyessenov commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,233 @@
+        "not supported]: ",
+        configuration->toString()));
+    incrementMetric(config_errors_, 1);
+    return false;
```
it'll crash immediately at this point.

kyessenov commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,233 @@
+      // __error attribute should be used by downstream plugins to distinguish
+      // between
+      // 1. No conditions matched.
+      // 2. Error evaluating a condition.
```
Why does it matter to downstream plugins? Either way, there is no value so the outcome is the same.

bianpengyuan replied:
+1. Can you elaborate some cases that you think it would be useful?

douglas-reid replied:
checking for errors is often easier than remembering to reason about absence of values, imho. any policy-like extensions relying on attribute-gen may only want to operate on a full set. is there another way to signal error in generation?

kyessenov commented:
Mostly looks good apart from a few comments. Will wait for other to finish the review.

kyessenov commented on extensions/attributegen/plugin_test.cc:
```diff
@@ -0,0 +1,489 @@
+    ASSERT_EQ(fs->hasData<WasmState>(attribute), found)
+        << absl::StrCat(attribute, "=?", value);
+    if (found) {
+      ASSERT_EQ(fs->getDataReadOnly<WasmState>(attribute).value(), value)
```
Since 1.14.1 PR merged, you want to update to `"wasm." + attribute` here I think.

kyessenov commented:
/retest

mandarjog commented on extensions/attributegen/plugin.cc:
```diff
@@ -0,0 +1,233 @@
+    }
+    gen_.push_back(AttributeGenerator(
+        phase, attribute_gen_config.output_attribute(), std::move(matches)));
+    matches.clear();
```
I just checked the c++ standard, and it says the object is in a valid but unspecified state.
You may use any operation that does not require a pre-condition.
So `.clear()` will get it into a known good state.

bianpengyuan replied:
Ok, why not declare a new vector at beginning of loop? seems easier to follow than using the same vector repeatedly.

mandarjog commented:
@bianpengyuan @kyessenov @douglas-reid all comments are addressed and `__error` attribute is removed.

bianpengyuan commented:
Woo, this was merged fast.. Wanted to block it if you want to fix the `matches` vector or wait for someone else to take another look. Anyway :)
